core/miner/storage_fsm/impl/precommit_batcher_impl.cpp

```diff
@@ -0,0 +1,156 @@
+        api_(std::move(api)),
+        miner_address_(miner_address),
+        closest_cutoff_(closest_cutoff) {}
+
+  PreCommitBatcherImpl::PreCommitEntry::PreCommitEntry(
+      const TokenAmount &number, const SectorPreCommitInfo &info)
+      : deposit(number), precommit_info(info){};
+
+  outcome::result<std::shared_ptr<PreCommitBatcherImpl>>
+  PreCommitBatcherImpl::makeBatcher(
+      const Ticks &max_wait,
+      std::shared_ptr<FullNodeApi> api,
+      const std::shared_ptr<libp2p::protocol::Scheduler>& scheduler,
+      const Address &miner_address) {
+    std::shared_ptr<PreCommitBatcherImpl> batcher =
+        std::make_shared<PreCommitBatcherImpl>(
+            max_wait, std::move(api), miner_address, max_wait);
+
+    batcher->cutoff_start_ = std::chrono::system_clock::now();
+    batcher->logger_ = common::createLogger("batcher");
+    batcher->logger_->info("Bather have been started");
+    batcher->handle_ = scheduler->schedule(max_wait, [=]() {
+      auto maybe_result = batcher->sendBatch();
+      for (const auto &[key, cb] : batcher->callbacks_) {
+        cb(maybe_result);
+      }
+      batcher->callbacks_.clear();
+      batcher->handle_.reschedule(max_wait);
+    });
+    return batcher;
+  }
+
+  outcome::result<CID> PreCommitBatcherImpl::sendBatch() {
+    std::unique_lock<std::mutex> locker(mutex_, std::defer_lock);
+    // TODO(Elestrias): [FIL-398] goodFunds = mutualDeposit + MaxFee; -  for
+    // checking payable
+    CID message_cid;
+    if (batch_storage_.size() != 0) {
+      logger_->info("Sending procedure started");
+
+      OUTCOME_TRY(head, api_->ChainHead());
+      OUTCOME_TRY(minfo, api_->StateMinerInfo(miner_address_, head->key));
+
+      PreCommitBatch::Params params = {};
+
+      for (const auto &data : batch_storage_) {
+        mutual_deposit_ += data.second.deposit;
+        params.sectors.push_back(data.second.precommit_info);
+      }
+
+      OUTCOME_TRY(encodedParams, codec::cbor::encode(params));
+
+      OUTCOME_TRY(signed_message,
+                  api_->MpoolPushMessage(
+                      vm::message::UnsignedMessage(
+                          miner_address_,
+                          minfo.worker,  // TODO: handle worker
+                          0,
+                          mutual_deposit_,
+                          {},
+                          {},
+                          25,  // TODO (m.tagirov) Miner actor v5 PreCommitBatch
+                          MethodParams{encodedParams}),
+                      kPushNoSpec));
+
+      mutual_deposit_ = 0;
+      batch_storage_.clear();
+      message_cid = signed_message.getCid();
+      logger_->info("Sending procedure completed");
+      cutoff_start_ = std::chrono::system_clock::now();
+      return message_cid;
+    }
+    cutoff_start_ = std::chrono::system_clock::now();
+    return ERROR_TEXT("Empty Batcher");
+  }
+
+  void PreCommitBatcherImpl::forceSend() {
+    auto maybe_result = sendBatch();
```

Comment:
Check `maybe_result.has_error()`

Reply:
I decided not to check the error, just to add the outcome::result  object  to provided callbacks

Reply:
And then handle error in caller proccess

Reply:
I'm not sure that it is a good idea.
Callbacks must consider that parameter could have an error.
@ortyomka what do you think?
Also I don't see any test with non-void callback

Reply:
This like subscibtion, this error is important for those who signed up to send the batch. 
But we can duplicate the error to the one who called Force.
I think this the best option

Reply:
the caller of sendForce can check callback if he needs to it, otherwise I cant imagine a situation of usage this functional, maybe logs can be used just to inform?